File: app/admin_auth.py
from __future__ import annotations

from sqladmin.authentication import AuthenticationBackend
from starlette.requests import Request
from starlette.responses import RedirectResponse

from app.database import db_manager
from app import models
from app.config import SECRET_KEY


class AdminAuth(AuthenticationBackend):
    async def login(self, request: Request) -> bool:
        # sqladmin default login page isn't used because we use the portal's login
        # However, if someone hits /admin/login directly, we redirect to our portal login
        return True

    async def logout(self, request: Request) -> bool:
        request.session.clear()
        return True

    async def authenticate(self, request: Request) -> bool:
        from app.config import AUTH_ENABLED
        if not AUTH_ENABLED:
            return True

        uid = request.session.get("user_id")
        if not uid:
            return False

        db = db_manager.SessionLocal()
        try:
            user = db.get(models.PortalUser, int(uid))
            if user and user.active and user.role == models.UserRole.admin:
                return True
        except (TypeError, ValueError):
            pass
        finally:
            db.close()

        return False


authentication_backend = AdminAuth(secret_key=SECRET_KEY)

File: crm-portal-ts/src/admin_auth.ts
import { Request } from 'express'; // Importing the correct type of request object as express does not have one by default, but we can use it.
// Also import RedirectResponse and AuthenticationBackend interface if you want to define them later on or in a different file 
from app.database import db_manager;  
import { PortalUser } from 'app/models'; // Assuming models is the directory where your portal user model resides, replace with actual path as per project structure   
// Also assuming SECRET_KEY type and interface are defined elsewhere in code 
from app.config import AUTH_ENABLED;  
import { UserRole } from 'app/models'; // Assuming models has a enum for user roles if not, replace with actual enums or types as per project structure   
// Also assuming db manager is the correct type of session local and get method 
class AdminAuth implements AuthenticationBackend{  
     async login(request: Request): Promise<boolean> { // Assuming request object has a property called 'session' which we can use to store user id. If not, replace with actual properties or methods as per project structure   
         return true;  }// same for logout and authenticate  
     async logout(request: Request): Promise<boolean> { // Assuming request object has a property called 'session' which we can use to clear user session. If not, replace with actual properties or methods as per project structure   
         return true;  }// same for login and authenticate  
     async authenticate(request: Request): Promise<boolean> { // Assuming request object has a property called 'session' which we can use to get user id. If not, replace with actual properties or methods as per project structure   
         if (!AUTH_ENABLED) return true;  }// same for login and logout  
     }